### DISCUSSION TITLE
running on Laravel Vapor

### DISCUSSION THREAD

**Discussion by 1stevengrant** [spatie/browsershot#654] [Q&A]
Would I be right in saying that this wouldn't work with Laravel Vapor given that there's no server for Chrome to be installed on?

**Comment by freekmurze** [comment]
Indeed. You can run it on Lambda, but you should research yourself how you can provision it with the requirements of this package.

**Comment by MaartenTell** [comment]
> Would I be right in saying that this wouldn't work with Laravel Vapor given that there's no server for Chrome to be installed on?

Hi, did you find a way to pull it of?

**Comment by stefro** [comment]
I’m also curious if you’ve figured this out.

**Comment by mkantautas** [comment]
Curious here as well o/

**Comment by jszobody** [comment]
Trying to accomplish the same thing, so far unsuccessfully. Would be amazing to find a blog article with a Browsershot / Vapor walkthrough.

**Comment by spaceemotion** [comment]
I  got this to work at https://github.com/laravel/vapor-php-build/issues/34#issuecomment-663286688 but there is one major pain point to this: I had to modify the browser.js. Could Browsershot provide two js files to overwrite? One that launches chrome and another that controls it?

**Comment by valentinfily** [comment]
Has anyone managed to make it work on Laravel Vapor with docker ?
I was using Lambda layers pretty successfully until recently, but Vapor requires to move to Docker images for future updates.

Until now I was using these layers in my `vapor.yml` file :
```
layers:      
  - "arn:aws:lambda:eu-west-3:553035198032:layer:nodejs12:37"
  - "vapor:php-7.4"
  - "arn:aws:lambda:eu-west-3:764866452798:layer:chrome-aws-lambda:20"
```

My Docker file (basically trying to replicate the behaviour of layers) looks like this :

```
FROM laravelphp/vapor:php74

RUN apk update && apk add --no-cache curl unzip

WORKDIR /opt

RUN curl -sSL https://github.com/shelfio/chrome-aws-lambda-layer/raw/master/chrome_aws_lambda.zip --output chrome_aws_lambda.zip

RUN unzip chrome_aws_lambda.zip -d /opt

RUN rm chrome_aws_lambda.zip

RUN curl -sSL https://github.com/lambci/node-custom-lambda/raw/master/v12.x/layer.zip --output node-custom-lambda.zip

RUN unzip node-custom-lambda.zip -d /opt/node-custom-lambda

RUN cp -avr /opt/node-custom-lambda/bin /opt/bin

RUN rm node-custom-lambda.zip && rm -rf node-custom-lambda

WORKDIR /

COPY . /var/task
```

**Comment by spaceemotion** [comment]
Pretty sure you wouldn't need a chrome-aws-lambda setup when using the docker image variant at all.
chrome-aws-lambda only exists because of the 200mb(?) image limit of lambda - which for docker is 10g if I'm not mistaken...

You should be able to install chrome as usual and use browsershot like any other virtual server.

That being said; actually installing chrome and getting everything to run is a different matter entirely. There are many dependencies that you need that don't seem to exist in the amazonlinux system...

**Comment by valentinfily** [comment]
Thanks, will report back when I get something working

**Comment by juliancc** [comment]
This Is what I have in the docker file. 

```
FROM laravelphp/vapor:php80

COPY . /var/task

# Installs latest Chromium (89) package.
RUN apk add --no-cache \
      chromium \
      nss \
      freetype \
      harfbuzz \
      ca-certificates \
      ttf-freefont \
      nodejs

RUN apk add --update npm

# Tell Puppeteer to skip installing Chrome. We'll be using the installed package.
ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true \
    PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser


RUN npm install --global --unsafe-perm puppeteer

# Add user so we don't need --no-sandbox.
RUN addgroup -S pptruser && adduser -S -g pptruser pptruser \
    && mkdir -p /home/pptruser/Downloads /app \
    && chown -R pptruser:pptruser /home/pptruser \
    && chown -R pptruser:pptruser /app

# Run everything after as non-privileged user.
USER pptruser

```

But when I run the following
```php
$url = 'https://www.google.com';

$screenshot = Browsershot::url($url)
            ->noSandbox()
            ->setScreenshotType('png')
            ->windowSize(1080,1920)
            ->screenshot();
```


I get:

```
Error: Protocol error (Target.setDiscoverTargets): Target closed.
   at /usr/local/lib/node_modules/puppeteer/lib/cjs/puppeteer/common/Connection.js:71:63
```

**Comment by argonzalez2016** [comment]
> This Is what I have in the docker file.
> 
> ```
> FROM laravelphp/vapor:php80
> 
> COPY . /var/task
> 
> # Installs latest Chromium (89) package.
> RUN apk add --no-cache \
>       chromium \
>       nss \
>       freetype \
>       harfbuzz \
>       ca-certificates \
>       ttf-freefont \
>       nodejs
> 
> RUN apk add --update npm
> 
> # Tell Puppeteer to skip installing Chrome. We'll be using the installed package.
> ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true \
>     PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser
> 
> 
> RUN npm install --global --unsafe-perm puppeteer
> 
> # Add user so we don't need --no-sandbox.
> RUN addgroup -S pptruser && adduser -S -g pptruser pptruser \
>     && mkdir -p /home/pptruser/Downloads /app \
>     && chown -R pptruser:pptruser /home/pptruser \
>     && chown -R pptruser:pptruser /app
> 
> # Run everything after as non-privileged user.
> USER pptruser
> ```
> 
> But when I run the following
> 
> ```
> $url = 'https://www.google.com';
> 
> $screenshot = Browsershot::url($url)
>             ->noSandbox()
>             ->setScreenshotType('png')
>             ->windowSize(1080,1920)
>             ->screenshot();
> ```
> 
> I get:
> 
> ```
> Error: Protocol error (Target.setDiscoverTargets): Target closed.
>    at /usr/local/lib/node_modules/puppeteer/lib/cjs/puppeteer/common/Connection.js:71:63
> ```

I'm in the same exact boat. Any help here would be awesome.

**Comment by theblindfrog** [comment]
I've managed to get a bit further.. but still haven't been able to crack it.

My dockerfile:
```docker
FROM laravelphp/vapor:php80

RUN apk update && \
    apk add --no-cache \
    nodejs \
    npm \
    chromium \
    harfbuzz \
    "freetype>2.8" \
    ttf-freefont \
    ca-certificates \
    nss

ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true \
    PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser

RUN npm install --global --unsafe-perm puppeteer

COPY . /var/task
```

My code:
```php
$html = "<HTML>... "; // etc

$pdfData = Browsershot::html($html)
            ->setNodeBinary('/usr/bin/node')
            ->setNpmBinary('/usr/bin/npm')
            ->setNodeModulePath('/usr/local/lib/node_modules')
            ->setChromePath("usr/local/lib/node_modules/puppeteer")
            ->addChromiumArguments([
                '--disable-web-security', '--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage'
            ])
            ->waitUntilNetworkIdle()
            ->emulateMedia('screen')
            ->showBackground()
            ->format('A4')
            ->setOption('args', ['--disable-web-security', '--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage'])
            ->pdf();
```
The 2nd `setOption` may be superfluous.

This setup doesn't throw an error... rather just returns a blank string (rather than the PDF data).
__I'm still to finish debugging why__ (as this is for a side project)... but if I crack it, I'll let folks know

I've ignored the whole 'running as a non-privileged user' aspect for the moment, to try and reduce the number of issues that could be affecting this.

**Comment by akhan619** [comment]
Same issue as above. Works like a charm on my Ubuntu 20.04 on WSL2. But get a blank pdf once deployed on Vapor. I thought maybe the issue is the version of chromium available on Alpine linux .

Alpine linux has version ```91.0.4472.164-r0``` for 3.14.  So tried matching them as best as I can by using ```puppeteer@9.1.1```  which depends on ```91.0.4469.0```. No luck.  

Side note: Tried to test the Alpine Linux environment locally by setting it up on WSL2. Sadly that gets stuck on simple npm install. Not to mention developing on it is a pain if you switch to it from Ubuntu. Would love for the Laravel team to just have a env for Ubuntu. Ofcourse that's wishful thinking.

**Comment by theblindfrog** [comment]
I have got it working @akhan619, but inconsistently with caveats. Again - still debugging.

I've made 2 changes to my browsershot code
```php
...
->setNodeModulePath("/var/task/node_modules")
->setChromePath("/usr/bin/chromium-browser") // per the dockerfile
```
I'm also not ` rm -rf node_modules` in my `vapor.yml` file under the build step.

This **_does_** allow me to produce pdfs or images from browsershot; but it's very, very, very slow. 
9/10 requests timeout. Not sure why.

I'm planning to pull the vapor-build docker image from ECR to play around with it more. I suspect the image is getting too large.

Another lead is around fonts having issues loading, and thus causing Chromium to timeout... but I haven't dug into that too deeply yet.

The other angle I'm thinking of trying is to get `chrome-aws-lambda` (per some of the originally success with layers) built into the docker image... but we'll see.

**Comment by juliancc** [comment]
@theblindfrog  can you share your docker file? 

don't understand what you mean by this
> i'm also not rm -rf node_modules in my vapor.yml file under the build step.

You mean you are installing puppeteer via node? and allowing chromium to be installed via puppeteer?

**Comment by theblindfrog** [comment]
Sure @juliancc 

**Dockerfile**
```dockerfile
FROM laravelphp/vapor:php80

# Install Node, Chromium
RUN apk update && \
    apk add --no-cache \
    nodejs \
    npm \
    chromium \
    harfbuzz \
    freetype \
    ttf-freefont \
    ca-certificates \
    nss

ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true \
    PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser \
    CHROME_BIN=/usr/bin/chromium-browser \
    NODE_ENV=production \
    HOME=/var/task

# Install Puppeteer
RUN npm install --global --unsafe-perm puppeteer

COPY . /var/task
```

**vapor.yml** (cut down) 
```yml
...
environments:
    staging:
        ...
        runtime: docker
        build:
            - 'COMPOSER_MIRROR_PATH_REPOS=1 composer install'
            - 'php artisan event:cache'
            - 'npm ci && npm run dev'
```
  
That last line is by default `'npm ci && npm run dev && rm -rf node_modules'`. Not removing the `node_modules` folder **is not a good idea**... but I'm clutching at straws to get this working!  



**Application code**
```php
$html = "<html><h1>Hello World</h1></html>";

$pdfData = Browsershot::html($html)
            ->setNodeBinary('/usr/bin/node')
            ->setNpmBinary('/usr/bin/npm')
            ->setNodeModulePath('/var/task/node_modules')
            ->setChromePath("/usr/bin/chromium-browser")
            ->addChromiumArguments(['--disable-web-security', '--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage', '--disable-gpu', '--single-process', '--headless'])
            ->setBinPath('/var/task/vendor/spatie/browsershot/bin/browser.js')
            ->emulateMedia('screen')
            ->timeout(300)
            ->waitUntilNetworkIdle()
            ->showBackground()
            ->format('A4')
            ->pdf();
```

**Comment by akhan619** [comment]
@theblindfrog  I actually did try keeping the node_modules instead of the ```rm -rf``` but ran into the ``` Failed to launch the browser process! ``` error. Here's the surprising part I actually went with a project only install of puppeteer, not global. When I remove the node_modules folder I get a blank pdf. It should complain about puppeteer missing. When I remove the node_modules I get the above error.

But I see you have used a different setup. I will try your setup tomorrow. We have to get this running lol. The other options for html -> pdf are a nightmare. Tried dompdf but that's stuck with CSS 2.1. Snappy is better but again has issues with css. Weasyprint v53 didn't even run for me. 

This is the best at this point with full support for CSS. Hell I can design the pdf with Tailwind and it's fast and beautiful. Anyways will let you guys know of any update.

**Comment by theblindfrog** [comment]
100% @akhan619. I've managed to get a PDF generated - albeit inconsistantly... _so this is possible!_

**Comment by argonzalez2016** [comment]
I had to use --single-process to get puppeteer with puphpeteer working on laravel vapor. I also had to utilize at least 1024 MB of memory.

**Comment by theblindfrog** [comment]
@argonzalez2016 but you were able to capture a screenshot/pdf, while running on vapor? Was this in a docker container? 
Could you share your code / setup?

**Comment by argonzalez2016** [comment]
**dockerfile**
Some of the env variables might not be necessary

```
FROM laravelphp/vapor:php80

# Installs latest Chromium (89) package.
RUN echo "http://dl-cdn.alpinelinux.org/alpine/edge/main" > /etc/apk/repositories \
    && echo "http://dl-cdn.alpinelinux.org/alpine/edge/community" >> /etc/apk/repositories \
    && echo "http://dl-cdn.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories \
    && echo "http://dl-cdn.alpinelinux.org/alpine/v3.12/main" >> /etc/apk/repositories \
    && apk upgrade -U -a \
    && apk add --no-cache \
    libstdc++ \
    chromium \
    nss \
    freetype \
    harfbuzz \
    ca-certificates \
    ttf-freefont \
    font-noto-emoji \
    wqy-zenhei \
    nodejs \
    npm \
    yarn

# Tell Puppeteer to skip installing Chrome. We'll be using the installed package.
ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true \
    PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser \
    PUPPETEER_SKIP_DOWNLOAD=true \
    CHROME_BIN=/usr/bin/chromium-browser \
    CHROME_PATH=/usr/lib/chromium/

WORKDIR /var/task

RUN yarn add @nesk/puphpeteer --dev

COPY . /var/task
```

**vapor.yml**
Removing the node modules directory is fine because yarn installs puppeteer in the docker container.

```
environments:
    placeHolder:
        cache: placeHolder-cache
        storage: placeHolder-ownerfactory
        domain: placeHolder.ownerfactory.com
        memory: 2048
        timeout: 60
        cli-memory: 1024
        cli-timeout: 600
        queue-memory: 2048
        queue-timeout: 600
        runtime: docker
        build:
            - 'COMPOSER_MIRROR_PATH_REPOS=1 composer install --optimize-autoloader --no-dev'
            - 'php artisan event:cache'
            - 'php artisan route:cache'
            - 'php artisan view:cache'
            - 'php artisan icons:cache'
            - 'yarn install && yarn run prod && rm -rf node_modules'
        deploy:
            - 'php artisan database:create'
            - 'php artisan migrate'
            - 'php artisan db:seed --class=ProductionSeeder'
            - 'php artisan cache:clear'
```

I'm able to use puphpeteer to take screenshots and make pdfs with HTTP requests, on the CLI, and in QUEUES. Should be similar to browsershot.
```
$chrome_settings = [
    'executablePath' => '/usr/bin/chromium-browser',
    'args' => [
        '--single-process',
        '--no-sandbox',
        '--no-zygote',
        '--disable-gpu',
	'--disable-setuid-sandbox',
	'--disable-dev-shm-usage',
        '--disable-web-security',
        '--disable-software-rasterizer',
        '--disable-features=IsolateOrigins,site-per-process',
        '--use-gl=swiftshader',
    ]
];
```

**Comment by freekmurze** [comment]
I'm try to add better support for Vapor out of the box.

Has any of you have a Docker file + configuration code that lets Browsershot run reliably on Vapor?

**Comment by spaceemotion** [comment]
I am currently building my own Docker-based image for Lambda (so not using the "out of the box" configuration).

My set up is based on the original files (https://github.com/laravel/vapor-php-build/tree/master/php80) in terms of runtime files and such. However, I decided to use an alpine image as the base, as it seems to have had the least problems with all the dependencies for chromium to actually run.

In order to run browsershot as smooth as possible I had to:
1. extend the original class to store files under /tmp
2. Disable as many flags as I can (this might vary if you want to render screenshots of pages, I disabled all 3D rendering as I only do PDF generation of blade templates)
3. Install all dependencies in my docker image

The most notable (and recent) change was hardcoding the `node_modules` directory, as npm 8.6 has a problem with non-existent `$HOME` directories (see https://github.com/npm/cli/issues/4769 and https://github.com/npm/cli/issues/4838). 

Not actually calling `npm` saved me about 400ms - on the 2gb lambda variant.

The full source code of the files can be found here:
https://gist.github.com/spaceemotion/01a9c94519eaf57fecba4279a8d83cc2

**Comment by alejandrozepeda** [comment]
@freekmurze I was able to run it in Vapor with this Dockerfile and configuration: 

```
FROM laravelphp/vapor:php81

RUN apk update && \
    apk add --no-cache \
    nodejs \
    npm \
    chromium \
    harfbuzz \
    freetype \
    ttf-freefont \
    ca-certificates \
    nss

ENV PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=true \
    PUPPETEER_EXECUTABLE_PATH=/usr/bin/chromium-browser

RUN npm install -g puppeteer

COPY . /var/task
```

Probably all the chromium arguments are not needed but I just imitate what chrome aws lambda does https://github.com/alixaxel/chrome-aws-lambda/blob/master/source/index.ts#L94
```
$screenshot = Browsershot::url(route('web.spaces.og-image.show', ['space' => $this->username]))
    ->setNodeModulePath('/usr/local/lib/node_modules')
    ->addChromiumArguments([
        'allow-running-insecure-content',
        'autoplay-policy' => 'user-gesture-required',
        'disable-component-update',
        'disable-domain-reliability',
        'disable-features' => 'AudioServiceOutOfProcess,IsolateOrigins,site-per-process',
        'disable-print-preview',
        'disable-setuid-sandbox',
        'disable-site-isolation-trials',
        'disable-speech-api',
        'disable-web-security',
        'disk-cache-size' => 33554432,
        'enable-features' => 'SharedArrayBuffer',
        'hide-scrollbars',
        'ignore-gpu-blocklist',
        'in-process-gpu',
        'mute-audio',
        'no-default-browser-check',
        'no-pings',
        'no-sandbox',
        'no-zygote',
        'use-gl' => 'swiftshader',
        'window-size' => '1920,1080',
        'single-process',
        'disable-2d-canvas-clip-aa',
        'disable-accelerated-2d-canvas',
        'disable-breakpad',
        'disable-canvas-aa',
        'disable-desktop-notifications',
        'disable-extensions',
        'disable-gl-drawing-for-tests',
        'disable-permissions-api',
        'disable-sync',
        'no-first-run',
    ])
    ->waitUntilNetworkIdle()
    ->ignoreHttpsErrors()
    ->disableJavascript()
    ->dismissDialogs()
    ->hideBrowserHeaderAndFooter()
    ->showBackground()
    ->setOption('landscape', true)
    ->windowSize(976, 549)
    ->setScreenshotType('jpeg', 100)
    ->screenshot();

Storage::disk(config('filesystems.public'))
    ->put($ogImagePath, $screenshot);
```

**Reply by ahinkle** [reply]
Thanks! Does chrome aws lambda not allow JavaScript (seeing it's disabled here)? When I remove that flag, it works for me - just making sure I'm not missing something. 🙂

**Reply by haoxi911** [reply]
I am testing the Chromium arguments in Laravel Sail, I have to remove the `'use-gl' => 'swiftshader'` line, otherwise the process is in a dead loop and never send back data.

**Comment by ahinkle** [comment]
Another great option is to use [Sidebar-Browsershot](https://github.com/stefanzweifel/sidecar-browsershot), which allows you to run Browsershot on AWS Lambda through [Sidecar](https://github.com/hammerstonedev/sidecar).

**Reply by bretto36** [reply]
This made it sooooo easy to do. Thanks @ahinkle
